nautobot_golden_config/utilities/helper.py

```diff
@@ -139,3 +150,103 @@ def get_device_to_settings_map(queryset):
+        secret_access_type (Optional[str], optional): Type of secret such as "Generic", "gNMI", "HTTP(S)". Defaults to "Generic".
+    .. rubric:: Example Jinja get_secret filter usage
+    .. highlight:: jinja
+    .. code-block:: jinja
+    password {{ id | get_secret("dcim.Device", "password") | encrypt_type5 }}
+    ppp pap sent-username {{ interface["connected_circuit_termination"]["circuit"]["id"] | get_secret("circuits.Circuit", "username") }} password {{ interface["connected_circuit_termination"]["circuit"]["id"] | get_secret("circuits.Circuit", "password") | encrypt_type7 }}
+
+    Returns:
+        Optional[str] : Secret value. None if there is no match. An error string if there is an error.
+    """
+    try:
+        module_name, model_name = obj_type.split(".")
+    except ValueError:  # not enough values to unpack (expected 2, got 1). We get this error if there is no "." in the obj_type
+        return f"Incorrect object type format {obj_type}. The correct format is 'module_name.model_name'."
+    try:
+        model = apps.get_model(module_name, model_name)
+    except ObjectDoesNotExist:
+        return f"Model {model_name} does not exist."
+    # get the object behind the model using the pk
+    try:
+        obj = model.objects.get(id=obj_id)
+    except ObjectDoesNotExist:
+        return f"Object wit ID {obj_id} does not exist."
+
+    # Get user permissions, terminate early if they do not have permission to view
+    app_label = obj._meta.app_label
+    model_name = obj._meta.model_name
+    permission_required = f"{app_label}.view_{model_name}"
+    # Bypass restriction for superusers and exempt views
+    if user.is_superuser or permission_is_exempt(permission_required):
+        pass
+    # User is anonymous or has not been granted the requisite permission
+    elif not user.is_authenticated or permission_required not in user.get_all_permissions():
+        return "You have no permission to read secrets. This incident will be reported."
+
+    # look through the attributes of the model to find if it has an id relationship
+    secrets_group = None
+    secrets_group_id = obj.__dict__.get("secrets_group_id")
+    if secrets_group_id:
+        secrets_group = SecretsGroup.objects.get(id=secrets_group_id)
+
+    # if the secret is not in the attributes, it may be in the relationships
+    for value in obj.get_relationships_data()["destination"].values():
+        # Find the relationship to secrets group
+        if isinstance(value["value"], SecretsGroup):
+            secrets_group = SecretsGroup.objects.get(slug=value["value"])
+            break
+
+    if secrets_group:
+        return secrets_group.get_secret_value(
+            access_type=secret_access_type,
+            secret_type=secret_type,
+            obj=obj,
+        )
+
+    return None
+
+
+def render_secrets(config_to_render: str, user: User):
+    """Renders secrets using the get_secrets filter, and Netutils ones."""
```

Comment:
To confirm, there is an idea that the raw will be stripped in first go around and in this go around, would be adding in the secret on the second?

Reply:
yes, this is the idea of this funciton. maybe should I add this comment here?